File: cpp/056_Merge_Intervals.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        if (intervals.size() == 0) return vector<vector<int>>();

        sort(intervals.begin(), intervals.end(), [](vector<int>& a, vector<int>& b) {
            return a[0] < b[0];
        });

        vector<vector<int>> res;

        for (int i = 1; i < intervals.size(); i++) {
            if (intervals[i-1][1] >= intervals[i][0]) {
                // overlap
                intervals[i][0] = intervals[i-1][0];
                intervals[i][1] = max(intervals[i-1][1], intervals[i][1]);
            } else {
                res.push_back(intervals[i-1]);
            }
        }
        res.push_back(intervals.back());

        return res;
    }
};

File: csharp/056_Merge_Intervals.cs
public class Solution {
    public int[][] Merge(int[][] intervals) {
        Array.Sort(intervals, (val1, val2) => val1[0].CompareTo(val2[0]));

        List<int[]> res = new List<int[]>();
        for (int i = 1; i < intervals.Length; i++) {
            if (intervals[i-1][1] >= intervals[i][0]) {
                intervals[i][0] = intervals[i-1][0];
                intervals[i][1] = Math.Max(intervals[i-1][1], intervals[i][1]);
            } else {
                res.Add(new int[] { intervals[i-1][0], intervals[i-1][1] });
            }
        }
        res.Add(new int[] { intervals[intervals.Length - 1][0], intervals[intervals.Length - 1][1] });

        return res.ToArray();
    }
}